Discord Rich Presence adaptation

doliinyk opened the pull request:
At the moment shows in your Discord game status engine name, location name or not if you in menu, spent time

doliinyk committed:
Поддержка Discord Rich Presence.
Показывает иконки, время игры, меню/локацию

doliinyk committed:
Поддержка Discord Rich Presence.
Показывает иконки, время игры, меню/локацию

collectioner commented on src/xrGame/MainMenu.cpp:
```diff
@@ -235,7 +237,9 @@ void CMainMenu::Activate(bool bActivate)
-        };
+        }
+        else
+            g_DiscordRPC->Update("Главное меню");
```
I think it'll be better to move this constant string to config file, since we support not only Russian language.
cc: @Xottab-DUTY what do you think?

Xottab-DUTY replied:
The game may even have that string translated (we need to find if that's true), so we can use it.

Generally, yes, we need to use translation, not just string hardcoded in the engine :)

collectioner commented on src/xrGame/GamePersistent.cpp:
```diff
@@ -612,6 +613,8 @@ extern CUISequencer* g_tutorial2;
 void CGamePersistent::OnFrame()
 {
+    
+
```
Please, remove these empty lines.

collectioner commented on src/xrGame/DiscordRPC.cpp:
```diff
@@ -0,0 +1,35 @@
+void DiscordRPC::Update(const char* level_name)
+{
+    DiscordRichPresence discordPresence;
+    memset(&discordPresence, 0, sizeof(discordPresence));
```
Not sure, but shouldn't we clear the memory after this variable usage?

Xottab-DUTY replied:
> Not sure, but shouldn't we clear the memory after this variable usage?

memset doesn't allocate memory.

collectioner commented on src/xrGame/GamePersistent.cpp:
```diff
@@ -938,6 +941,22 @@ void CGamePersistent::LoadTitle(bool change_tip, shared_str map_name)
+                temp_map_name = "Припять";
+            else if (map_name == "labx8")
+                temp_map_name = "Лаборатория X8";
+            g_DiscordRPC->Update(temp_map_name);
```
Again, that's fine for the first version. But I'd like to see these things in config files with localization. I'll leave priority of this for @Xottab-DUTY .

collectioner commented on src/xrGame/MainMenu.cpp:
```diff
@@ -203,6 +204,7 @@ void CMainMenu::Activate(bool bActivate)
 
     if (bActivate)
     {
+        g_DiscordRPC->Init();
```
Does it mean we initialize RPC every time we open main menu (well, when menu is active)?

Xottab-DUTY replied:
Yes. And it is totally wrong. Please, call this function in the constructor of CGamePersistent class, right before Ansel initialization.

collectioner commented on sdk/include/DiscordRPC/discord_register.h:
```diff
@@ -0,0 +1,26 @@
+#pragma once
```
@DenysOliinyk am I right that these 2 files (`discord_register.h` and `discord_rpc.`) are from Discord itself?

Xottab-DUTY replied:
@collectioner, yes, they are.

doliinyk replied:
@collectioner yup, from their official repo

collectioner replied:
Thanks you both, I resolved the comment.

Xottab-DUTY commented on src/xrGame/DiscordRPC.cpp:
```diff
@@ -0,0 +1,35 @@
+void DiscordRPC::Init()
+{
+    DiscordEventHandlers Handle;
+    memset(&Handle, 0, sizeof(Handle));
```
1. X-Ray uses ZeroMemory function for zeroing memory instead of memset and we use it too.
2. We don't even need manual memory zeroing. Just uses braces initializer;
3. First letter for local variable name should be lowercase. (we use `camelCase` for local variables. `snake_case` is allowed sometimes too)

```suggestion
    DiscordEventHandlers handle{};
```

Xottab-DUTY commented on src/xrGame/DiscordRPC.cpp:
```diff
@@ -0,0 +1,35 @@
+{
+    DiscordEventHandlers Handle;
+    memset(&Handle, 0, sizeof(Handle));
+    Discord_Initialize("869824972464488510", &Handle, 1, NULL);
```
1. Using NULL is outdated, it should be nullptr.
2. It's better to make a constant on top of the file (after includes) instead of hardcoded string with discord app ID.

Xottab-DUTY commented on src/xrGame/DiscordRPC.cpp:
```diff
@@ -0,0 +1,35 @@
+void DiscordRPC::Update(const char* level_name)
+{
+    DiscordRichPresence discordPresence;
+    memset(&discordPresence, 0, sizeof(discordPresence));
```
Again, no need for memset.
```suggestion
    DiscordRichPresence discordPresence{};
```

Xottab-DUTY commented on src/xrGame/DiscordRPC.h:
```diff
@@ -0,0 +1,17 @@
+#include <DiscordRPC/discord_register.h>
+#include <DiscordRPC/discord_rpc.h>
+#include <Windows.h>
+
```
Don't pay for what you don't use.
No need to include files which are not used. (include them in `.cpp`, except `windows.h`, it's already included in precompiled headers, globally)

And you should be especially careful with `windows.h` and not including it everywhere.
```suggestion
```

Xottab-DUTY commented on src/xrGame/DiscordRPC.h:
```diff
@@ -0,0 +1,17 @@
+    ~DiscordRPC();
+
+    void Init();
+    void Update(const char* level_name);
```
We use `pcstr` alias for `const char*`.
```suggestion
    void Update(pcstr level_name);
```

Xottab-DUTY commented on src/xrGame/DiscordRPC.cpp:
```diff
@@ -0,0 +1,35 @@
+DiscordRPC::~DiscordRPC()
+{
+    Discord_Shutdown();
+    Discord_ClearPresence();
```
Please, make the code symmetric: move deinitialization code to `Deinit` function.

doliinyk committed:
Update

doliinyk committed:
Update 1

Fixed all previous reviews, also now its translates task/location into game locales;

New: now shows active task;

Bugs: while loading save can not show active task, when loading save in another location task and location brokes, when loading save in current location task brokes and can be fixed by changing active task. Currently fine working on new game.

Xottab-DUTY commented on src/xrGame/GametaskManager.cpp:
```diff
@@ -238,7 +239,6 @@ CGameTask* CGameTaskManager::ActiveTask(ETaskType type)
         return nullptr;
 
     TASK_ID& t_id = g_active_task_id[type];
-
```
Return deleted string

Xottab-DUTY commented on src/xrGame/DiscordRPC.cpp:
```diff
@@ -0,0 +1,53 @@
+#include <xrCore/Text/StringConversion.hpp>
+
+DiscordRPC g_DiscordRPC;
+const pcstr g_discordAppID = "869824972464488510";
```
```suggestion
constexpr pcstr DISCORD_APP_ID = "869824972464488510";

DiscordRPC g_DiscordRPC;
```

Xottab-DUTY commented on src/xrGame/DiscordRPC.cpp:
```diff
@@ -0,0 +1,53 @@
+void DiscordRPC::Init()
+{
+    DiscordEventHandlers handle{};
+    Discord_Initialize(g_discordAppID, &handle, 1, nullptr);
```
```suggestion
    Discord_Initialize(DISCORD_APP_ID, &handle, 1, nullptr);
```

Xottab-DUTY commented on src/xrGame/DiscordRPC.cpp:
```diff
@@ -0,0 +1,53 @@
+    DiscordEventHandlers handle{};
+    Discord_Initialize(g_discordAppID, &handle, 1, nullptr);
+
+    xr_sprintf(build_name, "build %u", Core.GetBuildId());
```
Move it to the constructor

Xottab-DUTY replied:
I mean, move ONLY THIS LINE

Xottab-DUTY commented on src/xrGame/DiscordRPC.h:
```diff
@@ -0,0 +1,21 @@
+
+public:
+    DiscordRPC      (){ start_time = time(nullptr); };
+    ~DiscordRPC     (){};
```
Move constructor to .cpp
```suggestion
    DiscordRPC() { start_time = time(nullptr); }
    ~DiscordRPC();
```

Xottab-DUTY replied:
And make destructor default. There's no point in making it empty.

doliinyk replied:
Made `~DiscordRPC() = default;` 'cause compiler complains on `~DiscordRPC();`

Xottab-DUTY commented on src/xrGame/DiscordRPC.h:
```diff
@@ -0,0 +1,21 @@
+    void Init       ();
+    void Update     (pcstr  level_name = nullptr);
+    void SetTask    (pcstr  task_name);
+    void Deinit     ();
```
```suggestion
    void Init();
    void Deinit();

    void Update(pcstr level_name = nullptr);
    void SetTask(pcstr task_name);
```

doliinyk replied:
So I have to move Init() functions to constructor, may I delete function? There's no need of it else

Xottab-DUTY replied:
> So I have to move Init() functions to constructor

No

Xottab-DUTY replied:
Both constructor and Init stay the same, except there's just one line of code that should be moved from `Init` to constructor

Xottab-DUTY commented on src/xrGame/GamePersistent.cpp:
```diff
@@ -938,6 +941,9 @@ void CGamePersistent::LoadTitle(bool change_tip, shared_str map_name)
         pApp->LoadTitleInt(
             StringTable().translate("ls_header").c_str(), tmp.c_str(), StringTable().translate(buff).c_str());
+
+        if (map_name.c_str())
```
```suggestion
        if (!map_name.empty())
```

doliinyk replied:
![image](https://user-images.githubusercontent.com/68949838/127742052-6c46e4da-243f-4c55-b1b6-8b3d2c00ac52.png)
There's no empty() method for map_name

Xottab-DUTY replied:
Whoops, use this then:
```suggestion
        if (map_name.size())
```

Xottab-DUTY commented on src/xrGame/DiscordRPC.h:
```diff
@@ -0,0 +1,21 @@
+    int64_t     start_time          {};
+    string64    build_name          {};
+    string64    current_level_name  {};
+    pcstr       current_task_name   {};
```
Currently, we can make `current_level_name` to be local variable instead of member variable. :)
```suggestion
    int64_t start_time{};
    string64 build_name{};
    pcstr current_task_name{};
```

collectioner replied:
I agree. We can set all properties related to level inside `if` statement, when new level is passed.

doliinyk replied:
Did not make variable local, because then current_task_name overwrites location in status.

collectioner commented on src/xrGame/GametaskManager.cpp:
```diff
@@ -269,6 +270,7 @@ void CGameTaskManager::SetActiveTask(CGameTask* task)
         g_active_task_id[type] = task->m_ID;
         m_flags.set(eChanged, TRUE);
         task->m_read = true;
+        g_DiscordRPC.SetTask(StringTable().translate(task ? task->m_Title : "st_no_active_task").c_str());
```
I think, I'd like to move `StringTable().translate()` call inside `SetTask` and `Update` methods. My point is that if anything will be changed in the logic of getting localization, then we'll fix it only inside `DiscordRPC` class.
@Xottab-DUTY what do you think?

Xottab-DUTY replied:
I totally disagree.

Xottab-DUTY replied:
We should keep Discord integration abstract from localization system.

collectioner commented on src/xrGame/DiscordRPC.cpp:
```diff
@@ -0,0 +1,53 @@
+
+    discordPresence.state               = current_level_name;
+    discordPresence.largeImageText      = current_level_name;
+    discordPresence.details             = current_task_name;
```
@Xottab-DUTY should we leave such `column` style for assignment operators?

Xottab-DUTY replied:
Let's leave it as is, yes.

Xottab-DUTY commented on src/xrGame/DiscordRPC.cpp:
```diff
@@ -0,0 +1,53 @@
+#include "DiscordRPC.h"
+#include <DiscordRPC/discord_register.h>
+#include <DiscordRPC/discord_rpc.h>
+#include <xrCore/Text/StringConversion.hpp>
```
Using `<>` on own files is wrong, we should use quotes.
We also include third-party files after our own.
```suggestion
#include "DiscordRPC.h"
#include "xrCore/Text/StringConversion.hpp"

#include <DiscordRPC/discord_register.h>
#include <DiscordRPC/discord_rpc.h>
```

Xottab-DUTY commented on src/xrGame/DiscordRPC.h:
```diff
@@ -0,0 +1,22 @@
+
+class DiscordRPC
+{
+private:
```
Not needed, default access specifier is private already.
```suggestion
```

Xottab-DUTY commented on src/xrGame/GamePersistent.cpp:
```diff
@@ -38,6 +38,7 @@
 #include "xrCore/Threading/TaskManager.hpp"
 
 #include "xrPhysics/IPHWorld.h"
+#include "DiscordRPC.h"
```
Please, remove `AnselManager.h` inclusion on line 37 and make it like that:
```suggestion
#include "xrPhysics/IPHWorld.h"

#include "AnselManager.h"
#include "DiscordRPC.h"
```

Xottab-DUTY commented on src/xrGame/GamePersistent.cpp:
```diff
@@ -34,11 +34,12 @@
 #include "ui/UILoadingScreen.h"
-#include "AnselManager.h"
-#include "xrCore/Threading/TaskManager.hpp"
 
```
```suggestion
```

Xottab-DUTY commented on src/xrGame/DiscordRPC.h:
```diff
@@ -0,0 +1,21 @@
+public:
+    DiscordRPC();
+    ~DiscordRPC() = default;
+
```
```suggestion
```

Xottab-DUTY commented on src/xrGame/DiscordRPC.cpp:
```diff
@@ -0,0 +1,60 @@
+{
+    DiscordEventHandlers handle{};
+    Discord_Initialize(DISCORD_APP_ID, &handle, 1, nullptr);
+}
```
```suggestion
void DiscordRPC::Init()
{
    xr_sprintf(build_name, "build %u", Core.GetBuildId());
    start_time = time(nullptr);
    DiscordEventHandlers handle{};
    Discord_Initialize(DISCORD_APP_ID, &handle, 1, nullptr);
}
```

Xottab-DUTY reviewed (commented):
I just realized that `Core.GetBuildId()` can return 0 (at the point of executing constructor code), so it's better to move the code to `Init` function.

doliinyk committed:
Update 2
At the moment works fine, shows active location and task.

doliinyk committed:
Update 2
At the moment works fine, shows active location and task.

collectioner commented on src/xrGame/DiscordRPC.cpp:
```diff
@@ -0,0 +1,50 @@
+    discordPresence.largeImageKey   = "current_level_name";
+    discordPresence.smallImageKey   = "build_name";
+
+    //  type = TRUE  -> setting location
```
Please, remove comments.

Xottab-DUTY replied:
Why?

collectioner replied:
Well, yep, I see why do we need them now. But we can rename type to something more clear and then remove comments :)

collectioner commented on src/xrGame/DiscordRPC.cpp:
```diff
@@ -0,0 +1,50 @@
+    if (type)
+        xr_strcpy(current_level_name, temp.c_str());
+    else
+        xr_strcpy(current_task_name, temp.c_str());
```
Looks like it can be merged into conditional operator, like this:
`xr_strcpy(type ? current_level_name : current_task_name, temp.c_str());`

But I'd leave @Xottab-DUTY to decide If we need this adjustment.

Xottab-DUTY replied:
Using ternary operator is no longer relevant, since now we have enumeration instead of bool :)

collectioner commented on src/xrGame/DiscordRPC.h:
```diff
@@ -0,0 +1,17 @@
+    void Init();
+    void Deinit();
+
+    void Update(bool type, pcstr data);
```
I don't like the name of the first argument. It's not clear what type or how should we use it.
According to the usage of this argument, I'd suggest to rename it to smth like `isLevelUpdate` or similar.

cc: @Xottab-DUTY

collectioner replied:
Ok, let's update this as we discussed offline. I'll mark other related comments as resolved for now.

collectioner commented on src/xrGame/GametaskManager.cpp:
```diff
@@ -85,7 +86,11 @@ CGameTask* CGameTaskManager::HasGameTask(const TASK_ID& id, bool only_inprocess)
     if (it != GetGameTasks().end())
+    {
+        g_DiscordRPC.Update(
+            FALSE, StringTable().translate((*it).game_task ? (*it).game_task->m_Title : "st_no_active_task").c_str());
```
Since you declared first argument as bool, why do you use defines instead of standard bool variables, like `false` and `true`?

Xottab-DUTY commented on src/xrGame/DiscordRPC.h:
```diff
@@ -0,0 +1,23 @@
+    enum DiscordStatusType
+    {
+        UPDATE_LEVEL,
+        UPDATE_TASK
```
```suggestion
        UpdateLevel,
        UpdateTask,
```

Xottab-DUTY commented on src/xrGame/Level_start.cpp:
```diff
@@ -154,6 +155,9 @@ bool CLevel::net_start2()
         map_data.m_name = Server->level_name(m_caServerOptions);
+        if (map_data.m_name.size())
+            g_DiscordRPC.Update(
+                DiscordRPC::DiscordStatusType::UPDATE_LEVEL, StringTable().translate(map_data.m_name).c_str());
```
```suggestion
            g_DiscordRPC.Update(DiscordRPC::DiscordStatusType::UPDATE_LEVEL, StringTable().translate(map_data.m_name).c_str());
```

Xottab-DUTY commented on src/xrGame/DiscordRPC.cpp:
```diff
@@ -0,0 +1,47 @@
+    if (updateType == DiscordRPC::DiscordStatusType::UPDATE_LEVEL)
+        xr_strcpy(current_level_name, tempData.c_str());
+    else if (updateType == DiscordRPC::DiscordStatusType::UPDATE_TASK)
+        xr_strcpy(current_task_name, tempData.c_str());
```
```suggestion
    if (updateType == DiscordStatusType::UpdateLevel)
        xr_strcpy(current_level_name, tempData.c_str());
    else if (updateType == DiscordStatusType::UpdateTask)
        xr_strcpy(current_task_name, tempData.c_str());
```

Xottab-DUTY committed:
Apply suggestions from code review

doliinyk committed:
Revert "Update 2"

This reverts commit 475c556b63a26ddbb8f00f62d115e8f57aed506e.

doliinyk committed:
Merge branch 'rpc_integration' of https://github.com/protestant19/xray-16 into rpc_integration

doliinyk committed:
Update 3
Appeared dynamic libraries for debug configuration

doliinyk committed:
Fixed xrGame.vcxproj file: changed absolute path to dynamic library into relative

doliinyk committed:
Revert "Fixed xrGame.vcxproj file: changed absolute path to dynamic library into relative"

This reverts commit f75a50c93fb4197386d667b74658f1d6ff62d2f9.

doliinyk committed:
xrGame.vcxproj changed file copying tag

OPNA2608 commented on src/xrGame/GametaskManager.cpp:
```diff
@@ -85,7 +86,11 @@ CGameTask* CGameTaskManager::HasGameTask(const TASK_ID& id, bool only_inprocess)
     auto it = std::find_if(GetGameTasks().begin(), GetGameTasks().end(), key);
     if (it != GetGameTasks().end())
+    {
+        g_DiscordRPC.Update(DiscordRPC::DiscordStatusType::UpdateTask,
+            StringTable().translate((*it).game_task ? (*it).game_task->m_Title : "st_no_active_task").c_str());
```
Needs Windows-specific `#ifdef` blocks like in src/xrGame/GamePersistent.cpp:
```
CMakeFiles/xrGame.dir/GametaskManager.cpp.o:GametaskManager.cpp:function CGameTaskManager::HasGameTask(shared_str const&, bool): error: undefined reference to 'g_DiscordRPC'
CMakeFiles/xrGame.dir/GametaskManager.cpp.o:GametaskManager.cpp:function CGameTaskManager::HasGameTask(shared_str const&, bool): error: undefined reference to 'DiscordRPC::Update(DiscordRPC::DiscordStatusType, char const*)'
```

Xottab-DUTY replied:
No, no. No ifdefs, this can work on Linux, just needs to be fixed for it :)

OPNA2608 commented on src/xrGame/Level_start.cpp:
```diff
@@ -154,6 +155,8 @@ bool CLevel::net_start2()
         }
         Server->SLS_Default();
         map_data.m_name = Server->level_name(m_caServerOptions);
+        if (map_data.m_name.size())
+            g_DiscordRPC.Update(DiscordRPC::DiscordStatusType::UpdateLevel, StringTable().translate(map_data.m_name).c_str());
```
Needs Windows-specific `#ifdef` blocks like in src/xrGame/GamePersistent.cpp:
```
CMakeFiles/xrGame.dir/Level_start.cpp.o:Level_start.cpp:function CLevel::net_start2(): error: undefined reference to 'g_DiscordRPC'
CMakeFiles/xrGame.dir/Level_start.cpp.o:Level_start.cpp:function CLevel::net_start2(): error: undefined reference to 'DiscordRPC::Update(DiscordRPC::DiscordStatusType, char const*)'
```

Xottab-DUTY replied:
No, no. No ifdefs, this can work on Linux, just needs to be fixed for it :)

OPNA2608 reviewed (changes requested):
Breaks Linux builds in its current state.

doliinyk committed:
Fixed launching caused by wrong discord libraries name

doliinyk committed:
Updated deprecated Rich Presence SDK to newest Discord Game SDK

Xottab-DUTY committed:
Merge branch 'dev' into rpc_integration

doliinyk committed:
CMakeLists added new C++ file path

doliinyk committed:
Merge branch 'OpenXRay:dev' into rpc_integration

doliinyk committed:
Fix included stdafx.h into StdAfx.h

doliinyk committed:
Merge branch 'OpenXRay:dev' into rpc_integration
